### PR TITLE
Add support for remove services

### PR DESCRIPTION
**Pull request by pvizeli**
Add support for: https://github.com/home-assistant/home-assistant/pull/6441

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -45,6 +45,17 @@ export default function subscribeConfig(conn, configChanged) {
...
     }
 
+    function processServiceRemoved(event) {
+      if (config === null) return;
+
+      const { domain, service } = event.data;
+
+      if (!config.services[domain] || !config.services[domain][service]) return;
+
+      delete config.services[domain][service];
```
Comment (balloob) [inline lib/config.js]: - We should never mutate the existing object but always create a new one
 - Don't use `delete`. [It causes de-optimizations in V8](https://www.smashingmagazine.com/2012/11/writing-fast-memory-efficient-javascript/#de-referencing-misconceptions)

File: lib/config.js [hunk 2]
```diff
@@ -80,6 +91,7 @@ export default function subscribeConfig(conn, configChanged) {
     initPromise = Promise.all([
       conn.subscribeEvents(processComponentLoaded, 'component_loaded'),
       conn.subscribeEvents(processServiceRegistered, 'service_registered'),
+      conn.subscribeEvents(processServiceRemoved, 'service_removed'),
```
Comment (balloob) [inline lib/config.js]: This will return an unsub function that you should destructure in line 99 and call after line 103.

File: lib/config.js [hunk 3]
```diff
@@ -45,6 +45,14 @@ export default function subscribeConfig(conn, configChanged) {
       updateConfig({ services: Object.assign({}, config.services, { [domain]: domainInfo }) });
     }
 
+    function processServiceRemoved(event) {
+      if (config === null) return;
+
+      const { domain, service } = event.data;
+
+      updateConfig({ services: Object.assign({}, config.services, Object.assign({}, config.services.[domain], { [service]: null} )) });
```
Comment (balloob) [inline lib/config.js]: `config.services.[domain]` <- That's not valid JS.

Setting the service to `null` will break things too as people check for existence of key in this object. Instead, create your own copy object code that will skip the key.

```js
if (!(domain in config.services)) return;

const curDomainInfo = config.services[domain];

if (!curDomainInfo || !(service in curDomainInfo)) return;

const domainInfo = {};
Object.keys(curDomainInfo).forEach((sKey) => {
  if (sKey !== service) domainInfo[sKey] = curDomainInfo[sKey];
});
updateConfig({ services: Object.assign({}, config.services, { [domain]: domainInfo }) });
```

Comment (dale3h) [inline lib/config.js]: @balloob Would this not be more efficient?

```
function processServiceRemoved(event) {
  if (config === null) return;

  const { domain, service } = event.data;

  if (!config.services[domain] || !config.services[domain][service]) return;

  delete config.services[domain][service];
  if (!Object.keys(config.services[domain]).length)
    delete config.services[domain];

  updateConfig({});
}
```

Using `delete` does not set the value to `null`, but actually deletes the reference from memory.

Comment (balloob) [inline lib/config.js]: See https://github.com/home-assistant/home-assistant-js-websocket/pull/5#discussion_r104588121

**Pull request merged**